File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        int sum=0;
        int n = nums.size();
        vector<vector<int>> ans;

        for(int start=0;start<n;start++) {
            if(start>0 && nums[start]==nums[start-1]) { continue;}
            sum=nums[start];
            int i = start+1;
            int j = n-1;
            while(i<n && j>0 && i<j ) {
               // if(i>start+1 && nums[i]==nums[i-1]) { i++; continue;}
                if(sum+nums[i]+nums[j]==0) {
                    ans.push_back({nums[start],nums[i],nums[j]});
                    i++;
                    while(i<n && nums[i]==nums[i-1]) i++;
                } else if(sum+nums[i]+nums[j]>0) {
                    //We want to decrease sum
                    //Move j left
                    j-=1;
                } else {
                    i+=1;
                }
            }

        }
        return ans;
    }
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        i,j = 0,0
        nums.sort()
        n = len(nums)
        current_sum = 0
        answer = []
        for k  in range(n):
            if k>0 and nums[k]==nums[k-1]: continue
            i = k+1
            j = len(nums)-1
            while i<j and i<n and j<n:
                current_sum = nums[i]+nums[j]+nums[k]
                if current_sum<0:
                    i+=1
                elif current_sum>0:
                    j-=1
                else:
                    answer.append([nums[k],nums[i],nums[j]])
                    i+=1
                    while i>k+1 and i<n and nums[i]==nums[i-1]: i+=1
                #while j>=0 and j<n-1 and nums[j]==nums[j+1]: j-=1
        return answer

                
